File: source_unit.py
from __future__ import annotations

from pathlib import Path
from typing import List, Optional, Union

import processing.text_location
from errors.blame import BlameType, BlameSeverity
from processing.mode import ProcessingMode
from processing.options import ProcessingOptions
from utils import resolve_path


class SourceUnit:
    """Container of Axion source code.
    Different kinds of code processing
    are performed with that class.
    """

    def __init__(self, source_path: Path = None, output_path: Path = None, debug_path: Path = None):
        """
        Should not be used to instantiate class.
        """

        from processing.syntactic.expressions.ast import Ast
        from compiler import Compiler
        from errors.language_error import LanguageError
        from processing.syntactic.token_stream import TokenStream
        from processing.lexical.text_stream import TextStream

        self.options = ProcessingOptions.default
        self.mode = ProcessingMode.default

        self.text_stream: Optional[TextStream] = None

        from processing.lexical.tokens.token import Token, TokenType
        self.token_stream = TokenStream(self)
        self.mismatching_pairs: List[Token] = []
        self.process_terminators: List[TokenType] = [TokenType.end]

        self.indent_size = 0
        self.indent_char = '\0'
        self.last_indent_len = 0
        self.indent_level = 0

        self.blames: List[LanguageError] = []

        if source_path is None:
            source_path = Compiler.temp_source_path()
        self.source_path = resolve_path(source_path)

        if source_path.suffix != Compiler.source_file_ext:
            raise ValueError(f"Expected a '{Compiler.source_file_ext}' extension of '{source_path}' file.")

        if output_path is None:
            output_path = self.source_path.parent / "out" / (self.source_path.stem + Compiler.output_file_ext)
        self.output_path = resolve_path(output_path)

        if debug_path is None:
            debug_path = self.source_path.parent / "debug" / (self.source_path.stem + Compiler.debug_file_ext)
        self.debug_path = resolve_path(debug_path)

        self.ast = Ast(self)

    # region Constructors

    @staticmethod
    def from_code(source_code: str, out_file_path: Path = None) -> SourceUnit:
        """
        Initializes SourceUnit with source code.

        :param source_code: Axion source code to process.
        :param out_file_path: Path to file with compiler output.
        """
        from processing.lexical.text_stream import TextStream

        unit = SourceUnit(output_path = out_file_path)
        unit.text_stream = TextStream(source_code)
        return unit

    @staticmethod
    def from_lines(source_lines: List[str], out_file_path: Path = None) -> SourceUnit:
        """
        Initializes SourceUnit with source code lines.

        :param source_lines: Lines of Axion source code.
        :param out_file_path: Path to file with compiler output.
        """
        from processing.lexical.text_stream import TextStream

        unit = SourceUnit(output_path = out_file_path)
        unit.text_stream = TextStream("\n".join(source_lines))
        return unit

    @staticmethod
    def from_file(source_file_path: Path, out_file_path: Path = None) -> SourceUnit:
        """
        Initializes SourceUnit with source code from file.

        :param source_file_path: Path to file with source code.
        :param out_file_path: Path to file with compiler output.
        """
        from processing.lexical.text_stream import TextStream

        if not source_file_path.exists:
            raise FileNotFoundError(source_file_path)

        unit = SourceUnit(source_file_path, out_file_path)
        unit.text_stream = TextStream(source_file_path.read_text())
        return unit

    @staticmethod
    def from_interpolation(stream) -> SourceUnit:
        unit = SourceUnit()
        unit.text_stream = stream
        return unit

    # endregion

    def __repr__(self):
        return f"Source in '{self.source_path.name}' -> '{self.output_path.name}' [{self.debug_path.name}]"

    def blame(self, message: Union[str, BlameType], span: processing.text_location.Span,
              severity: BlameSeverity = None):
        from errors.language_error import LanguageError

        if isinstance(message, BlameType):
            if severity is None:
                severity = message.severity
            message = message.description
        self.blames.append(LanguageError(message, severity, span))

File: Axion.Core/Source/SourceUnit.cs
using System;
using System.Collections.Generic;
using System.IO;
using Axion.Core.Processing.CodeGen;
using Axion.Core.Processing.Errors;
using Axion.Core.Processing.Lexical;
using Axion.Core.Processing.Syntactic;
using Axion.Core.Processing.Syntactic.Expressions;
using Axion.Core.Specification;
using Newtonsoft.Json;

namespace Axion.Core.Source {
    /// <summary>
    ///     Container of Axion source code;
    ///     different kinds of code processing
    ///     are performed with that class.
    /// </summary>
    [JsonObject(MemberSerialization.OptIn)]
    public sealed class SourceUnit {
        public const string SourceFileExt = ".ax";
        public const string DebugFileExt  = ".dbg.json";

        public string OutputFileExt {
            get {
                if (Options.HasFlag(ProcessingOptions.ToAxion)) {
                    return ".ax";
                }

                if (Options.HasFlag(ProcessingOptions.ToCSharp)) {
                    return ".cs";
                }

                if (Options.HasFlag(ProcessingOptions.ToPython)) {
                    return ".py";
                }

                if (Options.HasFlag(ProcessingOptions.ToPascal)) {
                    return ".pas";
                }

                return ".out";
            }
        }

        internal ProcessingOptions Options        = ProcessingOptions.None;
        internal ProcessingMode    ProcessingMode = ProcessingMode.None;

        public List<LangException> Blames      { get; } = new List<LangException>();
        public TextStream          TextStream  { get; private set; }
        public TokenStream         TokenStream { get; }

        [JsonProperty]
        public Ast Ast { get; set; }

        private HashSet<string> CustomKeywords { get; } = new HashSet<string>();

        public CodeWriter CodeWriter { get; set; }

        private Dictionary<string, SourceUnit> Dependencies { get; } = new Dictionary<string, SourceUnit>();

        private Dictionary<string, IDefinitionExpr> Definitions { get; } = new Dictionary<string, IDefinitionExpr>();

        #region File paths

        /// <summary>
        ///     Path to file where source code is located.
        ///     When not specified in constructor,
        ///     then file name assigned to date and time of instance creation.
        /// </summary>
        public FileInfo SourceFilePath { get; }

        private FileInfo? outputFilePath;

        /// <summary>
        ///     Path to file where generated result is located.
        ///     When not specified in constructor,
        ///     then file name assigned to date and time of instance creation.
        /// </summary>
        public FileInfo? OutputFilePath {
            get {
                if (outputFilePath != null) {
                    return outputFilePath;
                }

                outputFilePath = new FileInfo(
                    Path.Combine(
                        SourceFilePath.Directory.FullName,
                        "out",
                        Path.GetFileNameWithoutExtension(SourceFilePath.FullName) + OutputFileExt
                    )
                );
                Utilities.ResolvePath(outputFilePath.Directory.FullName);
                return outputFilePath;
            }
        }

        private FileInfo? debugFilePath;

        /// <summary>
        ///     Path to file where processing debug output is located.
        /// </summary>
        public FileInfo DebugFilePath {
            get {
                if (debugFilePath != null) {
                    return debugFilePath;
                }

                debugFilePath = new FileInfo(
                    Path.Combine(
                        SourceFilePath.Directory.FullName,
                        "debug",
                        Path.GetFileNameWithoutExtension(SourceFilePath.FullName) + DebugFileExt
                    )
                );
                Utilities.ResolvePath(debugFilePath.Directory.FullName);
                return debugFilePath;
            }
        }

        #endregion

        private SourceUnit(
            string   code       = "",
            FileInfo sourcePath = null,
            FileInfo outputPath = null,
            FileInfo debugPath  = null
        ) {
            if (sourcePath == null) {
                sourcePath = new FileInfo(
                    Path.Combine(
                        Compiler.WorkDir,
                        "temp",
                        "tmp-" + DateTime.Now.ToFileName() + SourceFileExt
                    )
                );
                Utilities.ResolvePath(sourcePath.Directory?.FullName);
            }

            SourceFilePath = sourcePath;
            if (SourceFilePath.Extension != SourceFileExt) {
                throw new ArgumentException(
                    $"Source file must have {SourceFileExt} extension.",
                    nameof(sourcePath)
                );
            }

            if (outputPath != null) {
                outputFilePath = outputPath;
            }

            if (debugPath != null) {
                debugFilePath = debugPath;
            }

            TextStream  = new TextStream(code);
            TokenStream = new TokenStream();
            Ast         = new Ast(this);

            var macrosFile = new FileInfo(Path.Combine(Compiler.WorkDir, "modules", "macros.ax"));
            if (SourceFilePath.FullName != macrosFile.FullName) {
                AddDependency(FromFile(macrosFile));
            }
        }

        public void AddDependency(SourceUnit src) {
            Dependencies.Add(src.SourceFilePath.ToString(), src);
            Compiler.Process(src, ProcessingMode.Reduction);
        }

        public Dictionary<string, IDefinitionExpr> GetAllDefinitions() {
            foreach (SourceUnit dep in Dependencies.Values) {
                foreach (IDefinitionExpr def in dep.Ast.GetScopedDefs()) {
                    if (!Definitions.ContainsKey(def.Name.ToString())) {
                        Definitions.Add(def.Name.ToString(), def);
                    }
                }
            }
            return Definitions;
        }

        public void AddDefinition(IDefinitionExpr def) {
            if (def.Name != null) {
                if (Definitions.ContainsKey(def.Name.ToString())) {
                    LangException.Report(BlameType.NameIsAlreadyDefined, def.Name);
                }
                else {
                    Definitions.Add(def.Name.ToString(), def);
                }
            }
        }

        public HashSet<string> GetAllCustomKeywords() {
            foreach (SourceUnit dep in Dependencies.Values) {
                foreach (string kw in dep.CustomKeywords) {
                    CustomKeywords.Add(kw);
                }
            }
            return CustomKeywords;
        }

        public void RegisterCustomKeyword(string keyword) {
            if (!Spec.Keywords.ContainsKey(keyword)
             && !Spec.Operators.ContainsKey(keyword)
             && !Spec.Punctuation.ContainsKey(keyword)) {
                CustomKeywords.Add(keyword);
            }
        }

        public static SourceUnit FromCode(string code, FileInfo outFilePath = null) {
            return new SourceUnit(
                code,
                outputPath: outFilePath
            );
        }

        public static SourceUnit FromLines(string[] lines, FileInfo outFilePath = null) {
            return new SourceUnit(
                string.Join("\n", lines),
                outputPath: outFilePath
            );
        }

        public static SourceUnit FromFile(FileInfo srcFilePath, FileInfo outFilePath = null) {
            if (!srcFilePath.Exists) {
                throw new FileNotFoundException();
            }

            return new SourceUnit(
                File.ReadAllText(srcFilePath.FullName),
                srcFilePath,
                outFilePath
            );
        }

        public static SourceUnit FromInterpolation(TextStream stream) {
            return new SourceUnit { TextStream = stream };
        }
    }
}